sycl/include/CL/sycl/property_list.hpp

```diff
@@ -8,303 +8,96 @@
-} // namespace queue
-
-class noinit : public detail::Prop<detail::PropKind::NoInit> {};
-
-} // namespace property
-
-#if __cplusplus > 201402L
-
-inline constexpr property::noinit noinit;
-
-#else
-
-namespace {
-
-constexpr const auto &noinit =
-    sycl::detail::InlineVariableHelper<property::noinit>::value;
-
-}
-
-#endif
-
-class property_list {
-
-  // The structs validate that all objects passed are base of PropBase class.
-  template <typename... Tail> struct AllProperties : std::true_type {};
-  template <typename T, typename... Tail>
-  struct AllProperties<T, Tail...>
-      : std::conditional<std::is_base_of<property::detail::PropBase, T>::value,
-                         AllProperties<Tail...>, std::false_type>::type {};
-
-  template <class T>
-  using PropertyHolder = cl::sycl::property::detail::PropertyHolder<T>;
-  template <property::detail::PropKind PropKindT>
-  using Property = cl::sycl::property::detail::Prop<PropKindT>;
-
-  // The structs build tuple type that can hold all properties.
-  template <typename... Head> struct DefineTupleType {
-    using Type = std::tuple<Head...>;
-  };
-
-  template <int Counter, typename... Head>
-  struct BuildTupleType
-      : public std::conditional<
-            (Counter < property::detail::PropKind::PropKindSize),
-            BuildTupleType<
-                Counter + 1, Head...,
-                PropertyHolder<typename Property<(property::detail::PropKind)(
-                    Counter)>::FinalType>>,
-            DefineTupleType<Head...>>::type {};
-
-public:
-  // C'tor initialize m_PropList with properties passed by invoking ctorHelper
-  // recursively
-  template <typename... propertyTN,
-            typename = typename std::enable_if<
-                AllProperties<propertyTN...>::value>::type>
-  property_list(propertyTN... Props) {
-    ctorHelper(Props...);
-  }
-
-  template <typename propertyT> propertyT get_property() const {
-    static_assert((int)(propertyT::getKind()) <=
-                      property::detail::PropKind::PropKindSize,
-                  "Invalid option passed.");
-    const auto &PropHolder = std::get<(int)(propertyT::getKind())>(m_PropsList);
-    if (PropHolder.isInitialized()) {
-      return PropHolder.getProp();
-    }
-    throw invalid_object_error();
+  property_list() {}
+
+  // create a new property list from a parameter pack
+  template <
+      typename... propertyTN,
+      typename = typename std::enable_if<
+          are_base_of<property::detail::property_base, propertyTN...>::value,
+          void>::type>
+  property_list(propertyTN &&... args) {
+    properties.reserve(sizeof...(propertyTN));
+
+    init_property_list(std::forward<propertyTN>(args)...);
   }
 
+  // check if a property exists in the property list
   template <typename propertyT> bool has_property() const {
-    return std::get<(int)(propertyT::getKind())>(m_PropsList).isInitialized();
+    std::type_index index = std::type_index(typeid(propertyT));
```

Comment:
```suggestion
    using the_property = std::remove_cv<typename std::remove_referece<propertyT>::type>::type;
    std::type_index index = std::type_index(typeid(the_property));
```

Reply:
This may make it slightly more permissive, but I don't think this is needed. These functions need the property type to be explicitly stated, so the user has full control over whether or not it has any modifiers. The spec doesn't specify that it should be valid with modifiers however.

Note that the story is different for `init_property_list` as the template parameter `propertyT` is based on the properties passed to it, so it may get references.